src/response_decoder.ts

```diff
@@ -150,52 +153,121 @@ function decodeDeliverResponse(dataResponse: DataReader, logger: Logger): Messag
   const startFrom = dataResponse.position()
 
   let content = EmptyBuffer
+  let messageProperties: MessageProperties = {}
+  let applicationProperties: MessageApplicationProperties = {}
   while (dataResponse.position() - startFrom !== messageLength) {
     const formatCode = readFormatCodeType(dataResponse)
     switch (formatCode) {
       case FormatCodeType.ApplicationData:
         content = decodeApplicationData(dataResponse)
         break
-      case FormatCodeType.MessageAnnotations:
       case FormatCodeType.MessageProperties:
+        messageProperties = decodeMessageProperties(dataResponse)
+        break
       case FormatCodeType.ApplicationProperties:
+        applicationProperties = decodeApplicationProperties(dataResponse)
+        break
       case FormatCodeType.MessageHeader:
       case FormatCodeType.AmqpValue:
-        throw new Error("Not implemented")
         break
       default:
         throw new Error(`Not supported format code ${formatCode}`)
     }
   }
 
-  return content
+  return { content, messageProperties, applicationProperties }
+}
+
+function decodeApplicationProperties(dataResponse: DataReader) {
+  const formatCode = dataResponse.readUInt8()
+  const applicationPropertiesLength = decodeFormatCode(dataResponse, formatCode)
+  if (!applicationPropertiesLength) throw new Error(`invalid formatCode %#02x: ${formatCode}`)
+
+  return ApplicationProperties.parse(dataResponse, applicationPropertiesLength as number)
+}
+
+function decodeMessageProperties(dataResponse: DataReader) {
+  dataResponse.rewind(3)
+  const type = dataResponse.readInt8()
+  if (type !== 0) {
+    throw new Error(`invalid composite header %#02x: ${type}`)
+  }
+
+  const nextType = dataResponse.readInt8()
+  decodeFormatCode(dataResponse, nextType)
+
+  const formatCode = dataResponse.readUInt8()
+  const propertiesLength = decodeFormatCode(dataResponse, formatCode)
+  if (!propertiesLength) throw new Error(`invalid formatCode %#02x: ${formatCode}`)
+
+  return Properties.parse(dataResponse, propertiesLength as number)
 }
 
 function decodeApplicationData(dataResponse: DataReader) {
-  const type = dataResponse.readUInt8()
-  switch (type) {
+  const formatCode = dataResponse.readUInt8()
+  const length = decodeFormatCode(dataResponse, formatCode)
+  if (!length) throw new Error(`invalid formatCode %#02x: ${formatCode}`)
+
+  return dataResponse.readBufferOf(length as number)
+}
+
+function readFormatCodeType(dataResponse: DataReader) {
+  dataResponse.readUInt8()
+  dataResponse.readUInt8()
+
+  return dataResponse.readUInt8()
+}
+
+export function readUTF8String(dataResponse: DataReader) {
+  const formatCode = dataResponse.readUInt8()
+  const decodedString = decodeFormatCode(dataResponse, formatCode)
+  if (!decodedString) throw new Error(`invalid formatCode %#02x: ${formatCode}`)
+
+  return decodedString as string
+}
+
+function decodeFormatCode(dataResponse: DataReader, formatCode: number) {
```

Comment:
If you see, this method is very similar to `ApplicationProperties.decodeNextByteType`. Is it possible to create a single method?